File: scripts/check_balance.py
#!/usr/bin/env python3

import argparse
import requests
import os
import sys
from typing import Optional, Dict, Any
import structlog
import time
from datetime import datetime, timezone

logger = structlog.get_logger()

def check_balance(validator_url: str, wallet_address: str) -> Optional[float]:
    """Check BT2C wallet balance from a validator node"""
    try:
        response = requests.get(
            f"{validator_url}/blockchain/wallet/{wallet_address}/balance",
            headers={"Content-Type": "application/json"},
            timeout=5
        )
        if response.status_code == 200:
            data = response.json()
            return data.get("balance")
        else:
            logger.error("balance_check_failed",
                        status_code=response.status_code,
                        error=response.text)
            return None
    except Exception as e:
        logger.error("validator_connection_failed", error=str(e))
        return None

def check_validator_status(validator_url: str) -> Optional[Dict[str, Any]]:
    """Check validator node status and rewards"""
    try:
        response = requests.get(
            f"{validator_url}/blockchain/status",
            headers={"Content-Type": "application/json"},
            timeout=5
        )
        if response.status_code == 200:
            return response.json()
        else:
            logger.error("validator_status_check_failed",
                        status_code=response.status_code,
                        error=response.text)
            return None
    except Exception as e:
        logger.error("validator_connection_failed", error=str(e))
        return None

def check_staking_info(validator_url: str, wallet_address: str) -> Optional[Dict[str, Any]]:
    """Check staking information for a wallet"""
    try:
        response = requests.get(
            f"{validator_url}/blockchain/validator/{wallet_address}",
            headers={"Content-Type": "application/json"},
            timeout=5
        )
        if response.status_code == 200:
            return response.json()
        else:
            logger.error("staking_info_check_failed",
                        status_code=response.status_code,
                        error=response.text)
            return None
    except Exception as e:
        logger.error("validator_connection_failed", error=str(e))
        return None

def get_distribution_info(start_time: int) -> Dict[str, Any]:
    """Calculate distribution period information"""
    now = int(time.time())
    period_seconds = 14 * 24 * 60 * 60  # 14 days in seconds
    remaining_seconds = max(0, (start_time + period_seconds) - now)
    remaining_days = remaining_seconds // (24 * 60 * 60)
    
    return {
        "active": remaining_seconds > 0,
        "days_remaining": remaining_days,
        "end_date": datetime.fromtimestamp(start_time + period_seconds, tz=timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    }

def main():
    parser = argparse.ArgumentParser(description="Check BT2C wallet balance and validator status")
    parser.add_argument("--validator", default="http://localhost:8081",
                      help="Validator node URL (default: http://localhost:8081)")
    parser.add_argument("--wallet",
                      help="Wallet address to check")
    
    args = parser.parse_args()
    
    if not args.wallet:
        print("Error: Wallet address is required")
        parser.print_help()
        sys.exit(1)
    
    # Check validator status first
    print(f"\nChecking validator status at {args.validator}...")
    status = check_validator_status(args.validator)
    if status:
        print("\nValidator Status:")
        print(f"Network: {status.get('network', 'bt2c-mainnet-1')}")
        print(f"Block Height: {status.get('block_height', 0)}")
        print(f"Total Stake: {status.get('total_stake', 0)} BT2C")
        print(f"Block Time: {status.get('block_time', 300)}s")
        print(f"Initial Supply: {status.get('initial_supply', 21.0)} BT2C")
        
        # Show distribution period info
        dist_info = get_distribution_info(1710374400)  # March 14, 2025 00:00:00 UTC
        if dist_info["active"]:
            print("\nDistribution Period:")
            print(f"Days Remaining: {dist_info['days_remaining']}")
            print(f"End Date: {dist_info['end_date']}")
            print(f"Early Validators: {status.get('early_validator_count', 0)}")
            print(f"Total Rewards Distributed: {status.get('total_rewards_distributed', 0)} BT2C")
    
    # Check wallet balance
    print(f"\nChecking balance for wallet {args.wallet}...")
    balance = check_balance(args.validator, args.wallet)
    if balance is not None:
        print(f"Balance: {balance} BT2C")
    
    # Check staking info
    print("\nChecking staking information...")
    staking = check_staking_info(args.validator, args.wallet)
    if staking:
        print("\nStaking Status:")
        print(f"Total Staked: {staking.get('staked_amount', 0)} BT2C")
        print(f"Rewards Earned: {staking.get('rewards_earned', 0)} BT2C")
        print(f"Validator Status: {staking.get('is_validator', False)}")
        if staking.get('is_validator'):
            print(f"Validator Type: {'Developer Node' if staking.get('is_developer_node') else 'Early Validator'}")
            print(f"Blocks Validated: {staking.get('blocks_validated', 0)}")
            print(f"Uptime: {staking.get('uptime', 0)}%")
    
if __name__ == "__main__":
    main()

File: scripts/check_balance.js
/**
 * BT2C Wallet Balance Checker
 * 
 * This script checks the balance of a BT2C wallet address by:
 * 1. Connecting to the local node's API
 * 2. Querying the account endpoint to get balance information
 */

const http = require('http');
const https = require('https');
const fs = require('fs');
const path = require('path');

// Parse command line arguments
const args = process.argv.slice(2);
const addressArg = args.find(arg => arg.startsWith('--address='));
const apiUrlArg = args.find(arg => arg.startsWith('--api='));

if (!addressArg) {
  console.error('Error: Wallet address is required.');
  console.error('Usage: node check_balance.js --address=bt2c_address [--api=http://localhost:3000]');
  process.exit(1);
}

// Extract arguments
const address = addressArg.split('=')[1];
const apiUrl = apiUrlArg ? apiUrlArg.split('=')[1] : 'http://localhost:3000';

console.log('=== BT2C Wallet Balance Checker ===');
console.log(`Address: ${address}`);
console.log(`API URL: ${apiUrl}`);
console.log('=====================================');

// Function to make HTTP request
function makeRequest(url) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http;
    
    const req = client.get(url, (res) => {
      let data = '';
      
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            const parsedData = JSON.parse(data);
            resolve(parsedData);
          } catch (error) {
            reject(new Error(`Failed to parse response: ${error.message}`));
          }
        } else {
          reject(new Error(`Request failed with status code ${res.statusCode}: ${data}`));
        }
      });
    });
    
    req.on('error', (error) => {
      reject(new Error(`Request error: ${error.message}`));
    });
    
    req.end();
  });
}

// Check if the wallet data exists locally first
const walletDir = path.join(process.env.HOME || process.env.USERPROFILE, '.bt2c', 'validators');
const walletFile = path.join(walletDir, `${address}.json`);

// Try to get balance from API
async function checkBalanceFromAPI() {
  try {
    console.log('\nChecking balance from API...');
    const accountUrl = `${apiUrl}/api/v1/accounts/${address}`;
    const accountData = await makeRequest(accountUrl);
    
    console.log('\n=== Wallet Information ===');
    console.log(`Address: ${accountData.address}`);
    console.log(`Balance: ${accountData.balance} BT2C`);
    console.log(`Nonce: ${accountData.nonce}`);
    console.log(`Transaction Count: ${accountData.transactionCount}`);
    
    if (accountData.isValidator) {
      console.log('\n=== Validator Information ===');
      console.log(`Is Validator: Yes`);
      console.log(`Stake: ${accountData.stake} BT2C`);
      console.log(`Last Active: ${new Date(accountData.lastActive).toLocaleString()}`);
      
      // Try to get validator details
      try {
        const validatorUrl = `${apiUrl}/api/v1/validators/status/${address}`;
        const validatorData = await makeRequest(validatorUrl);
        
        console.log(`State: ${validatorData.state}`);
        console.log(`Reputation: ${validatorData.reputation}`);
        console.log(`Blocks Produced: ${validatorData.blocksProduced}`);
        console.log(`Blocks Validated: ${validatorData.blocksValidated}`);
        console.log(`Blocks Missed: ${validatorData.blocksMissed}`);
        console.log(`Uptime: ${validatorData.uptime}%`);
        
        if (validatorData.performance) {
          console.log('\n=== Performance (Last 24h) ===');
          console.log(`Blocks Produced: ${validatorData.performance.last24Hours.blocksProduced}`);
          console.log(`Blocks Missed: ${validatorData.performance.last24Hours.blocksMissed}`);
        }
      } catch (error) {
        // Validator details not available, continue with what we have
      }
    }
    
    // Try to get rewards
    try {
      const rewardsUrl = `${apiUrl}/api/v1/accounts/${address}/rewards?limit=5`;
      const rewardsData = await makeRequest(rewardsUrl);
      
      if (rewardsData.rewards && rewardsData.rewards.length > 0) {
        console.log('\n=== Recent Rewards ===');
        rewardsData.rewards.forEach((reward, index) => {
          console.log(`${index + 1}. Block ${reward.blockHeight}: ${reward.amount} BT2C (${reward.type})`);
        });
        console.log(`Total Rewards: ${rewardsData.total}`);
      }
    } catch (error) {
      // Rewards not available, continue with what we have
    }
    
  } catch (error) {
    console.error(`\nAPI Error: ${error.message}`);
    checkLocalWalletData();
  }
}

// Try to get balance from local wallet file
function checkLocalWalletData() {
  console.log('\nChecking local wallet data...');
  
  if (fs.existsSync(walletFile)) {
    try {
      const walletData = JSON.parse(fs.readFileSync(walletFile, 'utf8'));
      
      console.log('\n=== Local Validator Data ===');
      console.log(`Address: ${walletData.address}`);
      console.log(`Stake: ${walletData.stake} BT2C`);
      console.log(`State: ${walletData.state}`);
      console.log(`Reputation: ${walletData.reputation}`);
      console.log(`Is First Validator: ${walletData.isFirstValidator}`);
      console.log(`Joined During Distribution: ${walletData.joinedDuringDistribution}`);
      console.log(`Distribution Reward Claimed: ${walletData.distributionRewardClaimed}`);
      
      // Note: This is only validator data, not the full wallet balance
      console.log('\nNote: This is validator data only. For full wallet balance, the API must be accessible.');
      
    } catch (error) {
      console.error(`\nError reading local wallet file: ${error.message}`);
    }
  } else {
    console.error('\nLocal wallet data not found.');
    console.log('To check your balance:');
    console.log('1. Make sure the BT2C node is running');
    console.log('2. Try again with the correct API URL: node check_balance.js --address=your_address --api=http://your_node_ip:port');
  }
}

// Start by checking API
checkBalanceFromAPI();
